From b8cb2a5a721bacc8b96f26203075ce53fd1abb7b Mon Sep 17 00:00:00 2001
From: Ky-Anh Huynh <kyanh@theslinux.org>
Date: Sun, 28 Apr 2013 13:33:42 +0700
Subject: [PATCH] makepkg accepts PACKAGE_FEATURE

This variable is used to control some feature of output filename,
and of the package information
---
 pacman/src/pacman-4.1.0/scripts/makepkg.sh.in | 34 ++++++++++++++++++---------
 1 file changed, 23 insertions(+), 11 deletions(-)

diff --git a/pacman/src/pacman-4.1.0/scripts/makepkg.sh.in b/pacman/src/pacman-4.1.0/scripts/makepkg.sh.in
index b1eb313..9a9ca35 100644
--- a/pacman/src/pacman-4.1.0/scripts/makepkg.sh.in
+++ b/pacman/src/pacman-4.1.0/scripts/makepkg.sh.in
@@ -91,6 +91,13 @@ FORCE_VER=""
 
 PACMAN_OPTS=
 
+# Return the feature string to used in filename
+if [[ -n "$PACKAGE_FEATURE" ]]; then
+	_FEATURE_STRING="-$PACKAGE_FEATURE"
+else
+	_FEATURE_STRING=""
+fi
+
 shopt -s extglob
 
 ### SUBROUTINES ###
@@ -139,7 +146,6 @@ trap_exit() {
 	kill "-$signal" "$$"
 }
 
-
 ##
 # Clean up function. Called automatically when the script exits.
 ##
@@ -176,7 +182,7 @@ clean_up() {
 
 			# clean up dangling symlinks to packages
 			for pkg in ${pkgname[@]}; do
-				for file in ${pkg}-*-*-*{${PKGEXT},${SRCEXT}}; do
+				for file in ${pkg}${_FEATURE_STRING}-*-*-*{${PKGEXT},${SRCEXT}}; do
 					if [[ -h $file && ! -e $file ]]; then
 						rm -f "$file"
 					fi
@@ -1797,7 +1803,7 @@ write_pkginfo() {
 	printf "# Generated by makepkg %s\n" "$makepkg_version"
 	(( INFAKEROOT )) && printf "# using %s\n" "$(fakeroot -v)"
 	printf "# %s\n" "$(LC_ALL=C date -u)"
-	printf "pkgname = %s\n" "$pkgname"
+	printf "pkgname = %s\n" "$pkgname${_FEATURE_STRING}"
 	(( SPLITPKG )) && printf "pkgbase = %s\n" "$pkgbase"
 	printf "pkgver = %s\n" "$(get_full_version)"
 	printf "pkgdesc = %s\n" "${pkgdesc//+([[:space:]])/ }"
@@ -1810,6 +1816,12 @@ write_pkginfo() {
 	mapfile -t provides < <(find_libprovides)
 	mapfile -t depends < <(find_libdepends)
 
+	if [[ -n "$PACKAGE_FEATURE" ]]; then
+		conflicts=("${conflicts[@]}" "$pkgname")
+		provides=("${provides[@]}" "$pkgname")
+		replaces=("${replaces[@]}" "$pkgname")
+	fi
+
 	[[ $license ]]      && printf "license = %s\n"     "${license[@]}"
 	[[ $replaces ]]     && printf "replaces = %s\n"    "${replaces[@]}"
 	[[ $groups ]]       && printf "group = %s\n"       "${groups[@]}"
@@ -1867,7 +1879,7 @@ create_package() {
 
 	# tar it up
 	local fullver=$(get_full_version)
-	local pkg_file="$PKGDEST/${pkgname}-${fullver}-${pkgarch}${PKGEXT}"
+	local pkg_file="$PKGDEST/${pkgname}${_FEATURE_STRING}-${fullver}-${pkgarch}${PKGEXT}"
 	local ret=0
 
 	[[ -f $pkg_file ]] && rm -f "$pkg_file"
@@ -2060,10 +2072,10 @@ install_package() {
 	for pkg in ${pkgname[@]}; do
 		fullver=$(get_full_version $pkg)
 		pkgarch=$(get_pkg_arch $pkg)
-		pkglist+=("$PKGDEST/${pkg}-${fullver}-${pkgarch}${PKGEXT}")
+		pkglist+=("$PKGDEST/${pkg}${_FEATURE_STRING}-${fullver}-${pkgarch}${PKGEXT}")
 
-		if [[ -f "$PKGDEST/${pkg}-debug-${fullver}-${pkgarch}${PKGEXT}" ]]; then
-			pkglist+=("$PKGDEST/${pkg}-debug-${fullver}-${pkgarch}${PKGEXT}")
+		if [[ -f "$PKGDEST/${pkg}${_FEATURE_STRING}-debug-${fullver}-${pkgarch}${PKGEXT}" ]]; then
+			pkglist+=("$PKGDEST/${pkg}${_FEATURE_STRING}-debug-${fullver}-${pkgarch}${PKGEXT}")
 		fi
 	done
 
@@ -2351,7 +2363,7 @@ check_build_status() {
 	if (( ! SPLITPKG )); then
 		fullver=$(get_full_version)
 		pkgarch=$(get_pkg_arch)
-		if [[ -f $PKGDEST/${pkgname}-${fullver}-${pkgarch}${PKGEXT} ]] \
+		if [[ -f $PKGDEST/${pkgname}${_FEATURE_STRING}-${fullver}-${pkgarch}${PKGEXT} ]] \
 				 && ! (( FORCE || SOURCEONLY || NOBUILD )); then
 			if (( INSTALL )); then
 				warning "$(gettext "A package has already been built, installing existing package...")"
@@ -2368,7 +2380,7 @@ check_build_status() {
 		for pkg in ${pkgname[@]}; do
 			fullver=$(get_full_version $pkg)
 			pkgarch=$(get_pkg_arch $pkg)
-			if [[ -f $PKGDEST/${pkg}-${fullver}-${pkgarch}${PKGEXT} ]]; then
+			if [[ -f $PKGDEST/${pkg}${_FEATURE_STRING}-${fullver}-${pkgarch}${PKGEXT} ]]; then
 				somepkgbuilt=1
 			else
 				allpkgbuilt=0
@@ -2892,7 +2904,7 @@ if (( INFAKEROOT )); then
 fi
 
 fullver=$(get_full_version)
-msg "$(gettext "Making package: %s")" "$pkgbase $fullver ($(date))"
+msg "$(gettext "Making package: %s")" "$pkgbase${_FEATURE_STRING} $fullver ($(date))"
 
 if (( !PKGFUNC && !SPLITPKG )); then
 	warning "$(gettext "Using a %s without a %s function is deprecated.")" "$BUILDSCRIPT" "package()"
@@ -3043,7 +3055,7 @@ else
 fi
 
 fullver=$(get_full_version)
-msg "$(gettext "Finished making: %s")" "$pkgbase $fullver ($(date))"
+msg "$(gettext "Finished making: %s")" "$pkgbase${_FEATURE_STRING} $fullver ($(date))"
 
 install_package
 
-- 
1.8.3

